### ISSUE TITLE
Negative KL Loss

### ISSUE THREAD

**Issue by jeethesh-pai** [cxliu0/KL-Loss-pytorch#1]
Thank you for porting the code to mmdetection style. 
Did you encounter negative loss while training ?  As per the mathematical explanation of KL Loss. KL Loss seems always non negative.

**Comment by cxliu0** [comment]
I haven't encountered negative loss during training, but alpha (i.e., the second item of Eq. 9/Eq. 10 in the paper) could be negative. Nevertheless, the first item of Eq. 9/Eq. 10 is large enough, which ensures that L_reg is always larger than 0.

**Comment by jeethesh-pai** [comment]
Thank you so much for your quick reply .
Yea, even i found that the alpha was producing negative values. But in my case the first term was not large enough to make the whole sum positive. But can you tell me where you implemented this. 

![grafik](https://user-images.githubusercontent.com/75034628/188465076-f38d4cf8-96d8-4fcd-9ecc-2c07dc3329f2.png)

I saw in your code that you implemented an L1 loss, but in the paper they actually compare the xg, xe values before moving on to MSE loss or L1 loss.

**Comment by cxliu0** [comment]
I did use L1 loss, because I found that L1 loss is better than smooth-l1 loss. But I didn't inspect the loss value of smooth-l1 loss. Maybe (x_g-x_e)^2 is much smaller than (|x_g-x_e|-1/2) during training, which leads to negative loss. BTW, you can find the implementation in [convfc_bbox_head_kl.py](https://github.com/cxliu0/KL-Loss-pytorch/blob/main/mmdet/models/roi_heads/bbox_heads/convfc_bbox_head_kl.py#L332)

**Comment by jeethesh-pai** [comment]
Okay thanks a lot for the info.

**Comment by jeethesh-pai** [comment]
Can you tell me if you were calculating loss after bbox_decoding or before decoding using the predicted offsets ?

**Comment by cxliu0** [comment]
I compute regression loss before box decoding, i.e., [convfc_bbox_head_kl.py#L307](https://github.com/cxliu0/KL-Loss-pytorch/blob/main/mmdet/models/roi_heads/bbox_heads/convfc_bbox_head_kl.py#L307) is not executed.
